tools/cephfs: add pg_files command

jcsp opened the pull request:
tools/cephfs: add pg_files command
    
    Sometimes users know that particular data pool PGs
    have been damaged, and they would like to scan
    their files to work out which ones might have
    been affected.
    
    Fixes: http://tracker.ceph.com/issues/17249
    Signed-off-by: John Spray <john.spray@redhat.com>

jcsp commented:
Checked this doesn't currently break anything in fs suite but it needs its own test too

gregsfortytwo commented on src/tools/cephfs/PgFiles.cc:
```diff
@@ -0,0 +1,192 @@
+  layout.stripe_count = l_stripe_count;
+  layout.object_size = l_object_size;
+  layout.pool_id = l_pool_id;
+  if (r != 0) {
```
This delayed erro check isn't kinds weird?

jcsp replied:
Indeed it is, fixed.

gregsfortytwo commented on src/tools/cephfs/PgFiles.cc:
```diff
@@ -0,0 +1,192 @@
+  }
+
+  struct dirent de;
+  while((r = ceph_readdir_r(cmount, dr, &de)) != 0) {;
```
Gratuitous semi-colon? Or is there some preventive thing going on?

jcsp replied:
Fixed

gregsfortytwo commented on src/tools/cephfs/PgFiles.cc:
```diff
@@ -0,0 +1,192 @@
+      continue;
+    }
+
+    target.m_seed = ceph_stable_mod(target.ps(), pg_num, pg_num_mask);
```
Are there no OSDMap or otherwise more "public" functions exposing this?

jcsp replied:
Nope, in existing use in Objecter it's not OSDMap that does this part

gregsfortytwo commented:
My main concern here is that it will get used in place of a proper fsck sequence. It doesn't reconstruct the back traces on head objects, for instance.

Not sure if we should just note that in the docs or if there's a more definite care we can put in place (automaticallly re-creating any lost heads we detect here?)

jcsp commented:
I've updated this, and added a 'danger' block to the docs telling people to remove files and replace them rather than overwriting in place.  Ultimately we should also be repairing backtraces from forward scrub probably.

(Tested by https://github.com/ceph/ceph-qa-suite/pull/1216)

jcsp commented:
New test passing here: http://pulpito.ceph.com/jspray-2016-10-20_14:39:41-fs-wip-jcsp-testing-20161020-distro-basic-mira/487477/

jcsp commented:
@gregsfortytwo this should be good to go (and https://github.com/ceph/ceph/pull/11578 was inspired by this case too in case that wasn't obvious :-))
